File: 2015/7/go/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Circuit struct {
	wireInstructions map[string][]string
	wireCache        map[string]uint16
}

func NewCircuit() *Circuit {
	return &Circuit{
		wireInstructions: make(map[string][]string),
		wireCache:        make(map[string]uint16),
	}
}

func (c *Circuit) AddInstruction(instruction string) error {
	parts := strings.Split(instruction, " -> ")
	if len(parts) != 2 {
		return fmt.Errorf("invalid instruction format: %s", instruction)
	}

	targetWire := strings.TrimSpace(parts[1])
	source := strings.Fields(parts[0])
	c.wireInstructions[targetWire] = source
	return nil
}

func (c *Circuit) OverrideWire(wire string, value uint16) {
	c.wireInstructions[wire] = []string{strconv.Itoa(int(value))}
}

func (c *Circuit) ResetCache() {
	c.wireCache = make(map[string]uint16)
}

func (c *Circuit) GetWireValue(wire string) (uint16, error) {
	if val, ok := c.wireCache[wire]; ok {
		return val, nil
	}

	if val, err := strconv.Atoi(wire); err == nil {
		return uint16(val), nil
	}

	instruction, ok := c.wireInstructions[wire]
	if !ok {
		return 0, fmt.Errorf("wire not found: %s", wire)
	}

	var value uint16
	var err error

	switch len(instruction) {
	case 1:
		value, err = c.GetWireValue(instruction[0])

	case 2:
		if instruction[0] != "NOT" {
			return 0, fmt.Errorf("invalid unary operation: %v", instruction)
		}
		rightVal, err := c.GetWireValue(instruction[1])
		if err != nil {
			return 0, err
		}
		value = ^rightVal & 0xFFFF

	case 3:
		left, err := c.GetWireValue(instruction[0])
		if err != nil {
			return 0, err
		}
		right, err := c.GetWireValue(instruction[2])
		if err != nil {
			return 0, err
		}

		switch instruction[1] {
		case "AND":
			value = left & right
		case "OR":
			value = left | right
		case "LSHIFT":
			value = (left << right) & 0xFFFF
		case "RSHIFT":
			value = (left >> right) & 0xFFFF
		default:
			return 0, fmt.Errorf("invalid binary operation: %s", instruction[1])
		}

	default:
		return 0, fmt.Errorf("invalid instruction format: %v", instruction)
	}

	if err != nil {
		return 0, err
	}

	c.wireCache[wire] = value
	return value, nil
}

func solvePart1(instructions []string) (uint16, error) {
	circuit := NewCircuit()
	for _, instruction := range instructions {
		if err := circuit.AddInstruction(instruction); err != nil {
			return 0, err
		}
	}
	return circuit.GetWireValue("a")
}

func solvePart2(instructions []string, part1Result uint16) (uint16, error) {
	circuit := NewCircuit()
	for _, instruction := range instructions {
		if err := circuit.AddInstruction(instruction); err != nil {
			return 0, err
		}
	}
	circuit.OverrideWire("b", part1Result)
	return circuit.GetWireValue("a")
}

func main() {
	file, err := os.Open("../input.txt")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error opening file: %v\n", err)
		os.Exit(1)
	}
	defer file.Close()

	var instructions []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		instructions = append(instructions, scanner.Text())
	}

	part1Result, err := solvePart1(instructions)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Part 1 error: %v\n", err)
		os.Exit(1)
	}
	fmt.Printf("Part 1 - Wire 'a' value: %d\n", part1Result)

	part2Result, err := solvePart2(instructions, part1Result)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Part 2 error: %v\n", err)
		os.Exit(1)
	}
	fmt.Printf("Part 2 - Wire 'a' new value: %d\n", part2Result)
}

File: 2015/7/rust/src/main.rs
use std::collections::HashMap;
use std::fs;
use std::str::FromStr;

#[derive(Debug, Clone)]
enum Instruction{
    Direct(String),
    Not(String),
    And(String, String),
    Or(String, String),
    LShift(String, u16),
    RShift(String, u16),
}

struct Circuit {
    wire_instructions: HashMap<String, Instruction>,
    wire_cache: HashMap<String, u16>,
}

impl Circuit {
    fn new() -> Self {
        Self {
            wire_instructions: HashMap::new(),
            wire_cache: HashMap::new(),
        }
    }

    fn add_instruction(&mut self, instruction :&str) {
        let parts: Vec<&str> = instruction.split(" -> ").collect();
        if parts.len() != 2 {
            panic!("Invalid instruction format: {}", instruction);
        }
        let target_wire = parts[1].trim().to_string();
        let source = parts[0].trim();

        let parsed_instruction = match source.split_whitespace().collect::<Vec<&str>>().as_slice() {
            [val] => Instruction::Direct(val.to_string()),

            ["NOT", x] => Instruction::Not(x.to_string()),

            [x, "AND", y] => Instruction::And(x.to_string(), y.to_string()),

            [x, "OR", y] => Instruction::Or(x.to_string(), y.to_string()),

            [x, "LSHIFT", y] => Instruction::LShift(x.to_string(), y.parse().unwrap()),

            [x, "RSHIFT", y] => Instruction::RShift(x.to_string(), y.parse().unwrap()),

            _ => panic!("Invalid instruction format: {}", instruction),
        };
        self.wire_instructions.insert(target_wire, parsed_instruction);
    }

    fn override_wire(&mut self, wire: &str, value: u16) {
        self.wire_instructions.insert(wire.to_string(), Instruction::Direct(value.to_string()));
    }

    // fn reset_cache(&mut self){
    //     self.wire_cache.clear();
    // }

    fn get_wire_value(&mut self, wire: &str) -> u16 {
        if let Some(&cached_value) = self.wire_cache.get(wire) {
            return cached_value;
        }

        if let Ok(value) = u16::from_str(wire) {
            return value;
        }

        let instruction = self.wire_instructions.get(wire).expect("Wire not found").clone();
        let value = match instruction {
            Instruction::Direct(x) => self.get_wire_value(&x),

            Instruction::Not(x) => !self.get_wire_value(&x) & 0xFFFF,

            Instruction::And(x, y) => self.get_wire_value(&x) & self.get_wire_value(&y),

            Instruction::Or(x, y) => self.get_wire_value(&x) | self.get_wire_value(&y),

            Instruction::LShift(x, n) => (self.get_wire_value(&x) << n) & 0xFFFF,

            Instruction::RShift(x,n ) => self.get_wire_value(&x) >> n,
        };
        self.wire_cache.insert(wire.to_string(), value);
        value
    }
}

fn solve_part1(instructions: &[String]) -> u16 {
    let mut circuit = Circuit::new();
    for instruction in instructions{
        circuit.add_instruction(instruction);
    }
    circuit.get_wire_value("a")
}

fn solve_part2(instructions: &[String], part1_result: u16) -> u16 {
    let mut circuit = Circuit::new();
    for instruction in instructions{
        circuit.add_instruction(instruction);
    }
    circuit.override_wire("b", part1_result);
    circuit.get_wire_value("a")
}

fn main() {
    let data = fs::read_to_string("../input.txt").expect("Failed to read file");
    let instructions: Vec<String> = data.lines().map(|line| line.to_string()).collect();

    let part1_result = solve_part1(&instructions);
    println!("Part 1 - wire 'a' values: {}", part1_result);

    let part2_result = solve_part2(&instructions, part1_result);
    println!("Part 2 - wire 'a' values: {}", part2_result);
}